ASIM Authentication Event schema parser with its sample and test data for VMware Carbon Black Cloud

jayeshprajapaticrest opened the pull request:
Change(s):
   - Added ASIM Authentication Event schema parser for VMware Carbon Black Cloud based on API Data

   Reason for Change(s):
   - Initial version ASIM Authentication Event Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

v-rusraut commented:
Hi @jayeshprajapaticrest,
Thank you for raising Pull Request with us! We will review it and get back to you by shortly.

v-atulyadav commented:
Hi @jayeshprajapaticrest,
There is an error with the sample data and kql validation below. Please resolve it. Please check on branch conflicts as well. Thanks

<img width="619" alt="image" src="https://github.com/Azure/Azure-Sentinel/assets/104008048/19192a0e-1c05-43fd-bf4a-1a9a2985a4c8">

<img width="622" alt="image" src="https://github.com/Azure/Azure-Sentinel/assets/104008048/d1d3a7df-078a-41d7-a1fa-21c806e4325e">

<img width="599" alt="image" src="https://github.com/Azure/Azure-Sentinel/assets/104008048/e917f8de-03ba-472a-8696-0e65d7134534">

keyur-cds commented:
@v-atulyadav validations are fixed

vakohl commented on Parsers/ASimAuthentication/Parsers/vimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,86 @@
+          EventOwner = orgName_s
+      | extend DvcIpAddr = SrcIpAddr,
+          IpAddr = SrcIpAddr
+      | extend
```
can we merge these two extends?

niralishah-crest replied:
Merged as now we don't have DvcIpAddr field.

vakohl commented on Parsers/ASimAuthentication/Parsers/vimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,86 @@
+          SrcIpAddr = clientIp_s,
+          EventUid=_ItemId,
+          EventOwner = orgName_s
+      | extend DvcIpAddr = SrcIpAddr,
```
Is device where Authentication is performed and Device reporting the event are same? Is this product an EDR and event collected via endpoint agent?

niralishah-crest replied:
As we discussed for SentinelOne, similarly VMware Carbon Black Cloud also logs the public IP of network hence we have removed it from DvcIpAddr and considered it as SrcIpAddr only.

vakohl replied:
We should map DvcIpAddr based on following:
For authentication events, device fields refer to the system reporting the event.

niralishah-crest replied:
Yes. And that's why we haven't mapped DvcIpAddr fields with ip address coming from logs as it is not related to device.

vakohl commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_DataTest.csv:
```diff
@@ -0,0 +1,8 @@
+"(2) Info: Empty value in 191 records (95.98%)  in recommended field [EventResultDetails] (Schema:Authentication)"
+"(2) Info: Empty value in 196 records (98.49%)  in optional field [EventOriginalResultDetails] (Schema:Authentication)"
+"(2) Info: Empty value in 3 records (1.51%)  in recommended field [DvcIpAddr] (Schema:Authentication)"
+"(2) Info: Empty value in 3 records (1.51%)  in recommended field [SrcIpAddr] (Schema:Authentication)"
```
Why SrcIpAddr is coming null in some cases?
![image](https://github.com/Azure/Azure-Sentinel/assets/97222872/aac37a24-17ee-426f-b7c5-5dc06e769de9)

niralishah-crest replied:
We have few logs where clientIp_s field is empty which is mapped with SrcIpAddr.

vakohl commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(2) Info: Missing optional field [ThreatName]"
+"(2) Info: Missing optional field [ThreatOriginalConfidence]"
+"(2) Info: Missing optional field [ThreatOriginalRiskLevel]"
+"(2) Info: Missing optional field [ThreatRiskLevel]"
```
we have alot of missing fields, is this all we are getting in Authentication logs?

niralishah-crest replied:
Yes. there are very few fields available for Authentication logs.

vakohl replied:
we only using 5-6 fields from source

niralishah-crest replied:
Yes. As only that much fields are available from source itself.

niralishah-crest replied:
As we have only that much fields available from source.
![image](https://github.com/Azure/Azure-Sentinel/assets/107753570/03b82dba-53d2-4a4d-865c-241ce240ed12)

v-atulyadav commented:
Hi @jayeshprajapaticrest, could you please provide some updates on @vakohl's comments and also please look into branch conflicts. Thanks

jayeshprajapaticrest commented:
> Hi @jayeshprajapaticrest, could you please provide some updates on @vakohl's comments and also please look into branch conflicts. Thanks

@v-atulyadav Conflicts resolved and We will look at the comment by this weekend.

v-atulyadav commented:
Noted @jayeshprajapaticrest. Thanks

v-atulyadav commented:
Hi @jayeshprajapaticrest,
To follow up appropriately, we would appreciate some indication of the timeframe for these changes. Thanks

v-atulyadav commented:
Hi @jayeshprajapaticrest,
We're waiting for an update on this. Thanks

niralishah-crest commented on Parsers/ASimAuthentication/Parsers/imAuthentication.yaml:
```diff
@@ -67,4 +68,5 @@ Parsers:
   - vimAuthenticationCiscoISE
   - vimAuthenticationBarracudaWAF
   - vimAuthenticationVectraXDRAudit
+  - vimAuthenticationVMwareCarbonBlackCloud
```
Done

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(2) Info: Missing optional field [ThreatName]"
+"(2) Info: Missing optional field [ThreatOriginalConfidence]"
+"(2) Info: Missing optional field [ThreatOriginalRiskLevel]"
+"(2) Info: Missing optional field [ThreatRiskLevel]"
```
It's not part of the table schema. Previously, we have added _ItemId explicitly in raw logs but now we have removed it from raw logs as it will bw added by Sentinel itself.

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(2) Info: Missing optional field [TargetUserId]"
+"(2) Info: Missing optional field [TargetUserScopeId]"
+"(2) Info: Missing optional field [TargetUserScope]"
+"(2) Info: Missing optional field [ThreatCategory]"
```
Not available

vakohl replied:
@niralishah-crest We couldn't find logs that we could map with threat fields or the product itself doesn't support any threat info?

niralishah-crest commented on Sample Data/ASIM/VMware_CarbonBlackCloud_ASimAuthentication_RawLogs.json:
```diff
@@ -0,0 +1,338 @@
+[
```
We have only Logon type of events in VMware Carbon Black Cloud for Authentication.

vakohl replied:
@jayeshprajapaticrest @niralishah-crest 
If we had live environment, any specific reason we couldn't generated all event types?

niralishah-crest replied:
VMware Carbon Black supports only Login activities related to Authentication.
![image](https://github.com/Azure/Azure-Sentinel/assets/107753570/8ee5e1fb-7c08-485a-920d-b104a2892434)

vakohl replied:
@jayeshprajapaticrest @niralishah-crest 
are we sure there won't be any logout related logs coming VMWare side?

niralishah-crest replied:
In the official document of VMware, they have mentioned that login type of events are logged as part of AuditLogs and also in our live environment, only login attempts of users are logged as part of AuditLogs.

vakohl replied:
@niralishah-crest @jayeshprajapaticrest 
seems CarbonBlack does provide endpoint logs for authentication: https://developer.carbonblack.com/reference/carbon-black-cloud/cb-threathunter/latest/auth-events-api/#events-search
Current parser won't work for these logs.

niralishah-crest commented on Parsers/ASimAuthentication/Parsers/vimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,86 @@
+          EventSchemaVersion = "0.1.3",
+          EventVendor = "VMware",
+          EventType = "Logon",
+          EventResultDetails = iff(EventResult == "Failure", "Incorrect password", ""),
```
We have logs available only for Password did not match the type in case of failure.

vakohl replied:
@jayeshprajapaticrest @niralishah-crest 
any specific reason we didn't generated logs for other scenarios?

niralishah-crest replied:
As discussed, there is no specific type mentioned in the document that is supported by VMware Carbon Black. Still, our IT team has tried to perform different activities to generate logs.

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(1) Warning: Missing recommended field [DvcHostname]"
+"(1) Warning: Missing recommended field [TargetDomain]"
+"(1) Warning: Missing recommended field [TargetHostname]"
+"(2) Info: Missing optional alias [Application] aliasing non-existent column [TargetAppName]"
```
no

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(2) Info: Missing optional field [SrcDvcScope]"
+"(2) Info: Missing optional field [SrcFQDN]"
+"(2) Info: Missing optional field [SrcGeoCity]"
+"(2) Info: Missing optional field [SrcGeoCountry]"
```
no

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_ASimAuthentication_SchemaTest.csv:
```diff
@@ -0,0 +1,99 @@
+"(2) Info: Missing optional field [SrcGeoLatitude]"
+"(2) Info: Missing optional field [SrcGeoLongitude]"
+"(2) Info: Missing optional field [SrcGeoRegion]"
+"(2) Info: Missing optional field [SrcHostname]"
```
Not available

niralishah-crest commented on Parsers/ASimAuthentication/Tests/VMwareCarbonBackCloud_vimAuthentication_DataTest.csv:
```diff
@@ -0,0 +1,8 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 199 records (100.0%) for field [EventProduct] of type [Enumerated]: [""Carbon Black Cloud""] (Schema:Authentication)"
```
Done

niralishah-crest commented:
> Hi @jayeshprajapaticrest, We're waiting for an update on this. Thanks

@v-atulyadav We have made the suggested changes

v-atulyadav commented:
Noted @niralishah-crest. Thanks

vakohl commented on Parsers/ASimAuthentication/Parsers/imAuthentication.yaml:
```diff
@@ -50,6 +50,7 @@ ParserQuery: |
       , vimAuthenticationBarracudaWAF                   (starttime, endtime, targetusername_has, (imAuthenticationDisabled or('ExcludevimAuthenticationBarracudaWAF' in (DisabledParsers) )))
       , vimAuthenticationVectraXDRAudit               (starttime, endtime, (imAuthenticationDisabled or('ExcludevimAuthenticationVectraXDRAudit' in (DisabledParsers) )))
       , vimAuthenticationSentinelOne                    (starttime, endtime, (imAuthenticationDisabled or('ExcludevimAuthenticationSentinelOne' in (DisabledParsers) )))
+      , vimAuthenticationVMwareCarbonBlackCloud         (starttime, endtime, targetusername_has, (imAuthenticationDisabled or('ExcludevimAuthenticationVMwareCarbonBlackCloud' in (DisabledParsers) )))
```
Please update filters here

niralishah-crest replied:
Done

vakohl commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,73 @@
+          IpAddr = SrcIpAddr,
+          TargetUsernameType = _ASIM_GetUsernameType(TargetUsername),
+          TargetUserType = _ASIM_GetUserType(TargetUsername, ""),
+          Dvc = EventProduct,
```
If this is EDR product, isn't dvc be the system where agent is installed?

niralishah-crest replied:
Device information is not available in Authentication related logs.

vakohl commented on Parsers/ASimAuthentication/Parsers/vimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,125 @@
+    Type: bool
+    Default: false
+ParserQuery: |
+  let parser = (starttime: datetime=datetime(null), endtime: datetime=datetime(null), eventtype_in: dynamic=dynamic([]), eventresultdetails_in: dynamic=dynamic([]), eventresult: string='*', targetusername_has_any: string='*', targetappname_has_any: dynamic=dynamic([]), actorusername_has_any: string='*', srcipaddr_has_any_prefix: dynamic=dynamic([]), srchostname_has_any: dynamic=dynamic([]), targe ...
```
targetusername_has_any
actorusername_has_any
 both should be dynamic

niralishah-crest replied:
Updated

niralishah-crest replied:
Done

niralishah-crest commented on Parsers/ASimAuthentication/Parsers/vimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,125 @@
+      | extend
+          EventStartTime = unixtime_milliseconds_todatetime(eventTime_d),
+          AdditionalFields = bag_pack("flagged", flagged_b),
+          EventSeverity = iff(flagged_b == true, "Low", "Informational"),
```
flagged is true when there is a failure log for Authentication. So we have set severity as Low instead of Informational.

niralishah-crest commented:
@vakohl We have made the suggested changes.

jayeshprajapaticrest commented:
@vakohl 
We have updated unifying parser file with the filters as suggested by you. We have updated "targetusername_has(String Type)" with the "targetusername_has_any(Dynamic Type)" in unifying parser file. Also used the updated filter in the products developed by us. But the old products has "targetusername_has" filter name and because of that getting validation error as below.
**Errors:** The name 'targetusername_has' does not refer to any known column

vakohl commented:
> @vakohl We have updated unifying parser file with the filters as suggested by you. We have updated "targetusername_has(String Type)" with the "targetusername_has_any(Dynamic Type)" in unifying parser file. Also used the updated filter in the products developed by us. But the old products has "targetusername_has" filter name and because of that getting validation error as below. **Errors:** The name 'targetusername_has' does not refer to any known column

@jayeshprajapaticrest Please skip updating union for now.

v-atulyadav commented:
Hi @jayeshprajapaticrest, 
Please find the comments from @vakohl above and also check for errors in KQL validation. Thanks
<img width="905" alt="image" src="https://github.com/Azure/Azure-Sentinel/assets/104008048/91afaf73-cdca-4557-bf15-d427115e9861">

v-atulyadav commented:
Hi @jayeshprajapaticrest,
We're waiting for an update on this. Thanks

jayeshprajapaticrest commented:
> > @vakohl We have updated unifying parser file with the filters as suggested by you. We have updated "targetusername_has(String Type)" with the "targetusername_has_any(Dynamic Type)" in unifying parser file. Also used the updated filter in the products developed by us. But the old products has "targetusername_has" filter name and because of that getting validation error as below. **Errors:** The name 'targetusername_has' does not refer to any known column
> 
> @jayeshprajapaticrest Please skip updating union for now.

If we skip, still we can get the error as we have used "targetusername_has_any" of type dynamic. But previous version has "targetusername_has" of type string.

vakohl commented on Parsers/ASimAuthentication/Parsers/imAuthentication.yaml:
```diff
@@ -50,8 +80,9 @@ ParserQuery: |
       , vimAuthenticationBarracudaWAF                   (starttime, endtime, targetusername_has, (imAuthenticationDisabled or('ExcludevimAuthenticationBarracudaWAF' in (DisabledParsers) )))
       , vimAuthenticationVectraXDRAudit               (starttime, endtime, (imAuthenticationDisabled or('ExcludevimAuthenticationVectraXDRAudit' in (DisabledParsers) )))
       , vimAuthenticationSentinelOne                    (starttime, endtime, (imAuthenticationDisabled or('ExcludevimAuthenticationSentinelOne' in (DisabledParsers) )))
+      , vimAuthenticationVMwareCarbonBlackCloud         (starttime=starttime, endtime=endtime, eventtype_in=eventtype_in, eventresultdetails_in=eventresultdetails_in, eventresult=eventresult, targetusername_has_any=targetusername_has_any, targetappname_has_any=targetappname_has_any, actorusername_has_any=actorusername_has_any, srcipaddr_has_any_prefix=srcipaddr_has_any_prefix, srchostname_has_any ...
```
@jayeshprajapaticrest @niralishah-crest regatrding validation error. Can you please update this for existing parsers entires?

starttime=starttime, endtime=endtime, targetusername_has=targetusername_has_any, disabled=

niralishah-crest replied:
Done

v-atulyadav commented:
Hi @jayeshprajapaticrest, please check comments above from @vakohl. Thanks

v-rusraut commented:
Hi @jayeshprajapaticrest,
Please work on above comment.
Thanks

v-atulyadav commented:
Hi @jayeshprajapaticrest, could you please provide some updates on @vakohl's comments. Thanks

v-atulyadav commented:
Hi @jayeshprajapaticrest,
We're waiting for an update on this. Thanks

jayeshprajapaticrest commented:
@v-atulyadav 
We have started working on this PR suggested changes. Will update this by next week.
Thanks

v-atulyadav commented:
Noted @jayeshprajapaticrest. Thanks

vakohl commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,73 @@
+  let parser = (disabled: bool=false) {
+      CarbonBlackAuditLogs_CL
+      | where not(disabled)
+      | where description_s has_any ("logged in", "login") and description_s !has "connector"
```
@jayeshprajapaticrest @niralishah-crest 
found an additional description related to auth: "Failed at second factor authentication: the passcode doesn't match"

vakohl replied:
@jayeshprajapaticrest @niralishah-crest I'll take of this change.

vakohl commented on Parsers/ASimAuthentication/Parsers/ASimAuthenticationVMwareCarbonBlackCloud.yaml:
```diff
@@ -0,0 +1,73 @@
+          TargetUsername = loginName_s,
+          SrcIpAddr = clientIp_s,
+          EventUid=_ItemId,
+          EventOwner = orgName_s
```
this was mapped with ActorScope in Audit. In this authentication this could be mapped with TargetUserScope
![image](https://github.com/Azure/Azure-Sentinel/assets/97222872/01799bd1-7946-4708-bee2-9033d5b0b713)

vakohl replied:
@jayeshprajapaticrest @niralishah-crest will take care.

vakohl commented:
@jayeshprajapaticrest @niralishah-crest we are getting KQL validation error due to new Auth parameters. Can you remove the new changes in imauthentication union parser and only keep the old parameters to avoid the error for now?

vakohl commented:
> @jayeshprajapaticrest @niralishah-crest we are getting KQL validation error due to new Auth parameters. Can you remove the new changes in imauthentication union parser and only keep the old parameters to avoid the error for now?

@jayeshprajapaticrest @niralishah-crest I resolved remaining comments. Can you please fix the validation error?
